code/modules/mob/dead/new_player/latejoin_menu.dm

```diff
@@ -0,0 +1,192 @@
+		"departments" = departments,
+	)
+	if(SSshuttle.emergency)
+		switch(SSshuttle.emergency.mode)
+			if(SHUTTLE_ESCAPE)
+				data["shuttle_status"] = "The station has been evacuated."
+			if(SHUTTLE_CALL)
+				if(!SSshuttle.canRecall())
+					data["shuttle_status"] = "The station is currently undergoing evacuation procedures."
+
+	for(var/datum/job/prioritized_job in SSjob.prioritized_jobs)
+		if(prioritized_job.current_positions >= prioritized_job.total_positions)
+			SSjob.prioritized_jobs -= prioritized_job
+
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/list/department_jobs = list()
+		var/list/department_data = list(
+			"jobs" = department_jobs,
+			"open_slots" = 0,
+		)
+		departments[department.department_name] = department_data
+
+		for(var/datum/job/job_datum as anything in department.department_jobs)
+			var/job_availability = owner.IsJobUnavailable(job_datum.title, latejoin = TRUE)
+
+			var/list/job_data = list(
+				"prioritized" = (job_datum in SSjob.prioritized_jobs),
+				"used_slots" = job_datum.current_positions,
+				"open_slots" = job_datum.total_positions < 0 ? "∞" : job_datum.total_positions,
+			)
+
+			if(job_availability != JOB_AVAILABLE)
+				job_data["unavailable_reason"] = get_job_unavailable_error_message(job_availability, job_datum.title)
+
+			if(job_datum.total_positions < 0)
+				department_data["open_slots"] = "∞"
+
+			if(department_data["open_slots"] != "∞")
+				department_data["open_slots"] += job_datum.total_positions - job_datum.current_positions
+
+			department_jobs[job_datum.title] = job_data
+
+	return data
+
+/datum/latejoin_menu/ui_static_data(mob/user)
+	var/list/departments = list()
+
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/list/department_jobs = list()
+		var/list/department_data = list(
+			"jobs" = department_jobs,
+			"color" = department.ui_color,
+		)
+		departments[department.department_name] = department_data
+
+		for(var/datum/job/job_datum as anything in department.department_jobs)
+			var/datum/outfit/outfit = job_datum.outfit
+			var/datum/id_trim/trim = initial(outfit.id_trim)
+
+			var/list/job_data = list(
+				"command" = !!(job_datum.departments_bitflags & DEPARTMENT_BITFLAG_COMMAND),
+				"description" = job_datum.description,
+				"icon" = initial(trim.orbit_icon)
+			)
+
+			department_jobs[job_datum.title] = job_data
+
+	return list("departments_static" = departments)
+
+/datum/latejoin_menu/ui_status(mob/user)
+	return isnewplayer(user) ? UI_INTERACTIVE : UI_CLOSE
+
+/datum/latejoin_menu/ui_state(mob/user)
+	return GLOB.always_state
+
+/datum/latejoin_menu/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+
+	if(!ui.user.client || ui.user.client.interviewee || !isnewplayer(ui.user))
+		return TRUE
+
+	var/mob/dead/new_player/owner = ui.user
+
+	switch(action)
+		if("SelectedJob")
+			if(params["job"] == "Random")
+				var/list/dept_data = list()
```

Comment:
Can you split this body into its own proc? It makes this hard to read. Should just return the job since you pass it through.